lhotse/dataset/sampling/povey.py

```diff
@@ -0,0 +1,274 @@
+import math
+import os
+import random
+from pathlib import Path
+from typing import Dict, Generator, Iterable, List, Optional, Sequence, Tuple, Union
+
+import torch
+import torch.distributed as dist
+
+from lhotse import CutSet, Seconds
+from lhotse.cut.set import deserialize_cut
+from lhotse.dataset import DynamicBucketingSampler, DynamicCutSampler
+from lhotse.dataset.sampling.base import SamplingDiagnostics
+from lhotse.lazy import Dillable
+from lhotse.serialization import decode_json_line
+from lhotse.utils import Pathlike
+
+PathlikeAndScale = Tuple[Pathlike, float]
+
+
+class PoveySampler(torch.utils.data.Sampler, Dillable):
+    """
+    An infinite and stateless cut sampler that selects data at random from one or more cut manifests.
+    The main idea is to make training resumption easy while guaranteeing the data seen each
+    time by the model is shuffled differently.
+    It discards the notion of an "epoch" and it never finishes iteration.
+    It makes no strong guarantees about avoiding data duplication, but in practice you would
+    rarely see duplicated data.
+
+    This sampler's design was originally proposed by Dan Povey. For details see:
+    https://github.com/lhotse-speech/lhotse/issues/1096
+
+    Example 1: Get a bucketing :class:``.PoveySampler``::
+
+        >>> sampler = PoveySampler(
+        ...     cuts_paths=["data/cuts_a.jsonl", "data/cuts_b.jsonl"],
+        ...     index_path="data/files.idx",
+        ...     max_duration=600.0,
+        ...     num_buckets=50,
+        ...     quadratic_duration=30.0,
+        ... )
+
+    Example 1: Get a non-bucketing :class:``.PoveySampler``::
+
+        >>> sampler = PoveySampler(
+        ...     cuts_paths=["data/cuts_a.jsonl", "data/cuts_b.jsonl"],
+        ...     index_path="data/files.idx",
+        ...     max_duration=600.0,
+        ... )
+
+
+    .. note:: This sampler works only with uncompressed jsonl manifests, as it creates extra index files with line byte offsets to quickly find and sample JSON lines.
+     This means this sampler will not work with Webdataset and Lhotse Shar data format.
```

Comment:
Perhaps we should mention that if you restart training, you should set the random seed to some function of the step you restarted at, to ensure that you don't see the exact same data that you saw at the beginning of training (or at some previous restart point).

Reply:
Good point, I'll also add some logging that prints out the random seeds and maybe other info at initialization so it'll be possible to keep track of these things between experiments.